### ISSUE TITLE
add torch.stft and torch.fft

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#3775]
They would be super useful as fundamental ops (and autograd ops).

STFT can be implemented in terms of ConvTranspose, see github.com/pseeth/pytorch-stft
 

- [x] stft
- [x] istft
- [x] window functions: bartlett, hann, hamming, blackman
- [x] fft, ifft, irfft, rfft
- [x] fftshift, ifftshift

cc @mruberry

**Comment by PhilippPelz** [comment]
Don't know how you're feeling about this, but it would be nice to merge my branch with ZFloat and ZDouble types at some point. I have implemented fft, fft2, fft3, fftn, ifft, ifft2, ifft3, ifftn and fftshift, ifftshift already there. 

It would require some testing effort though as I have not written unit tests, but am actively using all the functions. 

https://github.com/PhilippPelz/pytorch


I will merge the upstream again when 0.3 comes out.

**Comment by SsnL** [comment]
looking into this

**Comment by SsnL** [comment]
I will implement `stft`. But for proper `fft`, I think we should use existing wheels, e.g., cuFFT and MKL/Eigen.

**Comment by PhilippPelz** [comment]
Yep, that's what I did. I surely didn't implement my own fft.

**Comment by diggerdu** [comment]
I have implemented STFT inverse STFT(trainable)
https://github.com/diggerdu/pytorch_audio

**Comment by soumith** [comment]
@diggerdu it's also implemented in master branch via #4095 
See: http://pytorch.org/docs/master/torch.html?highlight=stft#torch.stft

**Comment by diggerdu** [comment]
@soumith but without inverse transformation, which is an important module in google tacotron or other end to end model for audio processing. I have a paper which submitted to ICASSP 2018, we developed a  SOTA learning based model for noise reduction using this code. And a pure python implementation is more flexible and efficient than an unoptimized CUDA version, because we just combine some conv layers.

**Comment by soumith** [comment]
@diggerdu yea you're right, istft will be added soon too. 

> I have a paper which submitted to ICASSP 2018, we developed a SOTA learning based model for noise reduction using this code.

That's very cool. Are you planning to release the model training code as well?

**Comment by diggerdu** [comment]
@soumith 
Yeah, I will release that model after the accepted notification(or reject) been received. LOL

**Comment by rafaelvalle** [comment]
@diggerdu is that paper on arXiv?

**Comment by diggerdu** [comment]
@rafaelvalle be soon

**Comment by diggerdu** [comment]
@rafaelvalle Here is [sample](https://soundcloud.com/delphi-gentry/sets/the-end-to-end-model-for-speech-enhancement-by-jointly-masking)

**Comment by rafaelvalle** [comment]
@diggerdu Thanks for sharing this. Is there a channel, e.g. github or reddit, in which we can talk about your work? I don't want to spam this pytorch issue with an unrelated topic

**Comment by diggerdu** [comment]
@rafaelvalle [r/machinelearning](https://www.reddit.com/r/MachineLearning/comments/7phpfj/endtoend_model_for_speech_enhancement_by_joint/)
I'm glad you are interested in my first paper during undergraduate, and I can share more information 6h later because it's 3am, Shanghai.
All unrelated reply will be deleted after you received this.

**Comment by SsnL** [comment]
All fft methods are in master now. Once I add istft and (i)fftshift someday, this can be closed.

**Comment by rafaelvalle** [comment]
@SsnL please make sure it has windowing support, including corrections for windowing artifacts etc. For reference take a look at librosa's stft and isftf implementations:
https://librosa.github.io/librosa/_modules/librosa/core/spectrum.html#stft

**Comment by SsnL** [comment]
@rafaelvalle 

We have Hann, Hamming and Bartlett window functions available. Let us know if we forgot some important window functions.

By correction for windowing artifacts, do you mean that when used as a periodic window, the last value should be trimmed off?

Docs:
STFT: http://pytorch.org/docs/master/torch.html#torch.stft
Window fns: [Hann](http://pytorch.org/docs/master/torch.html#torch.hann_window), [Hamming](http://pytorch.org/docs/master/torch.html#torch.hamming_window), [Bartlett](http://pytorch.org/docs/master/torch.html#torch.bartlett_window)

**Comment by rafaelvalle** [comment]
There are modulation artifacts produced by windowing:
http://librosa.github.io/librosa/0.6.0rc0/generated/librosa.filters.window_sumsquare.html

The code they use to remove these artifacts lives here:
https://github.com/librosa/librosa/blob/6d791248832b45c0529045f3ed007860756b2a7a/librosa/core/spectrum.py#L302

If you add the blackman window, I think we have the standard set of windows supported by STFT implementations I have seen.

**Comment by SsnL** [comment]
Thanks for the pointers. I'll add this on the list of things to do.

I can add Blackman window. But the others might have to wait for after 0.4 because we are currently prioritizing polishing and releasing 0.4 first. :)

**Comment by keunwoochoi** [comment]
Is there any news on this issue? I'm interested in implementing STFT (and perhaps iSTFT) in pytorch too.

**Comment by SsnL** [comment]
Hi @keunwoochoi , stft is already in. I just need to add the unchecked methods.

**Comment by vinaykumar2491** [comment]
Hi @SsnL How can we use `torch.stft` on CUDA. I don't see any `device` parameter in `torch.stft()`

**Comment by tomrunia** [comment]
I have written `fftshift` and `ifftshift` operations as part of my Complex Steerable Pyramid implementation: https://github.com/tomrunia/PyTorchSteerablePyramid 

(see the `steerable/math_utils.py` file)

**Comment by SsnL** [comment]
@tomrunia Thanks! Your repo looks really cool, btw :).

**Comment by veritas9872** [comment]
Hello. I would like to request a feature in the current implementation of fft, ifft, etc. 

Currently, there is no control over the axes that the Fourier Transform is performed on. 
Only the number of dimensions can be specified. 

I believe that this means that the last 'signal_ndim' dimensions are transformed.

However, the documentation does not explicitly mention which dimensions are transformed. See [here](https://pytorch.org/docs/stable/torch.html#torch.fft) to see what I mean.

Although this may serve most purposes, some use cases involve controlling which axes are transformed. This is very easy in [numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.fft.fft.html) but impossible in the current Pytorch implementation.

Although intermediate axes can be transformed by first transforming all axes and then inverse transforming others, or by reordering the axes for the Fourier Transform and then returning them to their original order, both these methods are very inefficient. 

I would therefore like to request adding control over which axes are transformed.

**Comment by veritas9872** [comment]
After looking into the issue I found that [torch.transpose](https://pytorch.org/docs/stable/torch.html#torch.transpose) does not make a memory copy. So transposing the tensor a few times might be the best solution...

**Comment by veritas9872** [comment]
@PhilippPelz Could an optimized version of fftshift and ifftshift also be added? Since they require lots of data to be reshaped, it is quite possible that they are taking up as much time as the FFT operations themselves. Is there any way to implement the shift operations so as to minimize data transfer/copy?

**Comment by chen1399** [comment]
I have question when I try Pytorch for C++. 
Tensor stft( ) missing the first output point and the last output point.
For example:
The output's shape is (161,153) when I use torch_stft (python).
The output's shape is (161,151) when I use Tensor stft( ) (C++).
audio length is 24320
lenEnc is audio length
`
#python from torch_stft import STFT
stft = STFT(
    filter_length=320, 
    hop_length=160, 
    win_length=320,
    window='hamming'
).to('cpu')
magnitude, phase = stft.transform(audio)
output = stft.inverse(magnitude, phase)`
`
//C++
    at::Tensor wavTen = at::from_blob(wavNor,{lenEnc});
    at::Tensor hamm = at::hamming_window(320);
    at::Tensor spec = at::stft(wavTen,320,160,320,hamm,false,true);
`

**Comment by vincentqb** [comment]
> Does 'torch.istft' currently under development?

istft is available in torchaudio: [code](https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L57), [tests](https://github.com/pytorch/audio/blob/master/test/test_functional.py), [documentation](https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L57). Feel free to open PRs to migrate from torchaudio to pytorch.

**Comment by mruberry** [comment]
I believe this issue has finally been addressed now that https://github.com/pytorch/pytorch/pull/44877 is in!

**Comment by JXuann** [comment]
Thanks for all of your implementation efforts! Is it correctly understood that the window size of window functions is also trainable? For Hann/Hamming window, as the window size is positive int, how could it be made differentiable in your implementation? Many thanks in advance!

**Comment by mruberry** [comment]
> Thanks for all of your implementation efforts! Is it correctly understood that the window size of window functions is also trainable? For Hann/Hamming window, as the window size is positive int, how could it be made differentiable in your implementation? Many thanks in advance!

The window size is not immediately trainable, although you could implement your own method of updating it. It's an interesting idea.

**Comment by JXuann** [comment]
> > Thanks for all of your implementation efforts! Is it correctly understood that the window size of window functions is also trainable? For Hann/Hamming window, as the window size is positive int, how could it be made differentiable in your implementation? Many thanks in advance!
> 
> The window size is not immediately trainable, although you could implement your own method of updating it. It's an interesting idea.

Thanks for confirming it. I get stuck at finding a way to differentiate the Hann window function.
